### DISCUSSION TITLE
SQLAlchemy 2.0 and support for `os.PathLike`

### DISCUSSION THREAD

**Discussion by labrys** [sqlalchemy/sqlalchemy#9027] [Usage Questions] [answered]
Currently in SQLAlchemy 2.0.0b4 the following code `path: Mapped[os.PathLike]` results in `sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type: <class 'os.PathLike'>`.  As `os.PathLike` is an abstract class that resolves to a string or bytes I would expect it to be supported as a text type.

**Comment by CaselIT** [comment]
Hi,

I'm -1 on this. The type map is used customizable in multiple ways so that's what it should be used: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#orm-declarative-mapped-column-type-map

Also, there is no database type that maps both string or bytes, so a decision would need to be made by sqlalchemy on something that the user is in a better position to make

**Reply by labrys** [reply]
With a raw `os.PathLike` I would agree, but with `os.PathLike[str]` or `os.PathLike[bytes]` it would resolve to an explicit type.

**Reply by zzzeek** [reply]
not a bad point we dont map any sub-types like that by default either.

in any case, assuming String for the column type wont actually even work since your SQL expressions will be returning strings, not PathLike instances.

**Reply by peterjc** [reply]
A related question, can we do something like ``Mapped(pathlib.Path)``?

That fails for me with ``qlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type <class 'pathlib.Path'> inside the 'XXX' attribute Mapped annotation``

**Reply by CaselIT** [reply]
sure, you just need to add path to type map. Refer to https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#orm-declarative-mapped-column-type-map

**Reply by peterjc** [reply]
That's encouraging, but evidentially I'm still missing something:

```python
from pathlib import Path
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

class Base(DeclarativeBase):
    """Base class for SQLAlchemy ORM declarations."""
    type_annotation_map = {
        Path : String,  # Want to store pathlib.Path objects like strings
    }

class Paths(Base):
    __tablename__ = "paths"
    idn: Mapped[int] = mapped_column(primary_key=True)
    path: Mapped[Path] = mapped_column()

# Quick test
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
engine = create_engine(url="sqlite:///demo.db")
Base.metadata.create_all(engine)
session = sessionmaker(bind=engine)()
path_entry = Paths(path=Path("/tmp/"))
session.add(path_entry)
session.commit()
```

This fails at the commit with ``sqlite3.ProgrammingError: Error binding parameter 1: type 'PosixPath' is not supported``

If this can be done, it seems like it could make a good addition to the documentation.

**Reply by CaselIT** [reply]
You will likely need to to create a type decorator that transforms to string and vice versa. You can look at https://docs.sqlalchemy.org/en/20/core/custom_types.html#backend-agnostic-guid-type for an example

**Reply by peterjc** [reply]
Ah - I see now labrys' example for an ``os.PathLike`` mapping (which was hidden by default in the GitHub display of this discussion). That does the ``TypeDecorator`` magic. Thanks!

**Comment by zzzeek** [comment, marked as answer]
the annotation-based mapping system does not currently traverse ``__mro__`` for a given class, such as if you had string type `MyStr(str)`, it would not in that case resolve to String either unless you set up a mapping for it.

This is for among other reasons because using the String datatype, you will get str objects back, not `MyStr`.

looking at `os.PathLike` it has no hierarchy shared with `str` in any case:

```py
>>> import os
>>> os.PathLike.__mro__
(<class 'os.PathLike'>, <class 'abc.ABC'>, <class 'object'>)
```

"string" and "bytes" also refer to different SQL types.

therefore I would not expect SQLAlchemy to have a default SQL mapping for the ``os.PathLike`` type, and in any case, there's no SQLAlchemy SQL type that returns `PathLike` objects so there's no default mapping to be set up for this type.

As mentioned in the docs, any and all Python types you wish to map to SQL types should be set up front in the type map: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#orm-declarative-mapped-column-type-map

**Reply by labrys** [reply]
Makes sense, thanks for the quick response.

**Reply by labrys** [reply]
Does this look like an appropriate recipe for a PathLike mapping?

```python
import os
import pathlib
from typing import Callable
from typing import Optional

from sqlalchemy import Dialect
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import registry
from sqlalchemy.types import TypeDecorator


class PathLike(TypeDecorator):
    """Allows mapping an `os.PathLike` object to an `sqlalchemy.String`"""

    impl = String

    def __init__(self, factory: Callable[[str], os.PathLike]):
        super().__init__()
        self.factory = factory

    def process_bind_param(self, value: Optional[os.PathLike], dialect: Dialect) -> str:
        """Convert an `os.PathLike` value to a string for the database."""
        if value:
            value = os.fspath(value)
        return value

    def process_result_value(self, value: Optional[str], dialect: Dialect) -> Optional[os.PathLike]:
        """Restore a string from the database to an `os.Pathlike`."""
        if value is not None:
            value = self.factory(value)
        return value


class Base(DeclarativeBase):
    """Base class used for declarative class definitions."""

    registry = registry(
        type_annotation_map={
            os.PathLike: PathLike(pathlib.Path),
        }
    )


class SomeClass(Base):
    """A sample ORM class with a PathLike column."""

    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    path: Mapped[os.PathLike]
```

**Reply by zzzeek** [reply]
perfect!

**Reply by CaselIT** [reply]
Maybe it could be added to the wiki?

**Reply by zzzeek** [reply]
maybe we should just document using the type map with the ORM over at https://docs.sqlalchemy.org/en/20/core/custom_types.html#typedecorator-recipes

**Reply by labrys** [reply]
Documenting the type map usage (or at least linking to it) in the typedecorator-recipes would be an improvement.

If there is an appropriate section in the wiki, I would not be opposed to it being included there as well.

**Reply by labrys** [reply]
A tip or note in the relevant section of the tutorials would help with discoverability.  Something that states what types `Mapped` supports natively and a link to the appropriate documentation on how to customize the type-maps.  Maybe put the note in [quickstart.html#declare-models](https://docs.sqlalchemy.org/en/20/orm/quickstart.html#declare-models) for the ORM QuickStart and for the SQLAlchemy Unified Tutorial in [metadata.html#declaring-mapped-classes](https://docs.sqlalchemy.org/en/20/tutorial/metadata.html#declaring-mapped-classes).

**Reply by peterjc** [reply]
I think the last line of the example is wrong:

```python
    path: Mapped[PathLike]
```

That gives me ``sqlalchemy.exc.ArgumentError: The type provided inside the 'path' attribute Mapped annotation is the SQLAlchemy type <class '__main__.PathLike'>. Expected a Python type instead``

Either of the following work for me, first the abstract base class:

```python
    path: Mapped[os.PathLike]
```

Or the more specific concrete class:

```python
    path: Mapped[pathlib.Path]
```

I agree this would be a good addition to the documentation.

**Reply by CaselIT** [reply]
I don't think there is a PathLike global, so I guess it was imported from `os`

**Reply by peterjc** [reply]
The code as currently given defines a local class ``PathLike`` which subclasses ``TypeDecorator``. The name collision adds to the confusion.

**Reply by CaselIT** [reply]
You are right. I've edited the message to use the name from os, since in Mapped should go python types, not db types

**Comment by CaselIT** [comment]
I've added some updates to the documentation here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4348
